src/string_scanner.cr

```diff
@@ -125,6 +148,42 @@ class StringScanner
     end
   end
 
+  private def match(pattern : String, advance = true, anchored = true)
+    @last_match = nil
+    if pattern.bytesize > @str.bytesize - @byte_offset
+      nil
+    elsif anchored
+      if @str.unsafe_byte_slice(@byte_offset, pattern.bytesize) == pattern.unsafe_byte_slice(0)
+        @last_match = StringMatchData.new(pattern)
+        @byte_offset += pattern.bytesize if advance
+        pattern
+      end
+    elsif (found = @str.byte_index(pattern, @byte_offset))
+      finish = found + pattern.bytesize
+      result = @str.byte_slice(@byte_offset, finish - @byte_offset)
+      @byte_offset = finish if advance
+      @last_match = StringMatchData.new(result)
+      result
+    end
+  end
+
+  private def match(pattern : Char, advance = true, anchored = true)
+    @last_match = nil
+    if pattern.bytesize > @str.bytesize - @byte_offset
+      nil
+    elsif anchored
+      unsafe_str = String.new(@str.unsafe_byte_slice(@byte_offset))
+      if unsafe_str.starts_with?(pattern)
```

Comment:
This leads to quadratic heap memory usage if the same scanner keeps matching `Char`s, and should be avoided. The following is based on `Char#to_s(IO)`: (`i` is equal to `pattern.bytesize` at the end of the loop)

```suggestion
      char_bytes = uninitialized UInt8[4]
      i = 0
      pattern.each_byte do |byte|
        char_bytes[i] = byte
        i += 1
      end
      if @str.unsafe_byte_slice(@byte_offset, i) == char_bytes.to_slice[0, i]
```

Reply:
(Also this is not the most beautiful code, but there are ways to encapsulate this internally similar to `Crystal::SmallQueue`, so don't worry about its look)

Reply:
Alternatively, just compare byte by byte directly. Something like:

```crystal
i = 0
pattern.each_byte do |byte|
  return nil unless @str.to_unsafe[@byte_offset + i] == byte
  i += 1
end
# no if-block here
```

Reply:
@HertzDevil , thank you for suggestion.
Yep, I've understand `String.unsafe_byte_slice_string` wrong, so decided those construction wont allocate.

Used you version with per-byte comparison. + added `String.byte_index(Char)` for un-anchored search.